keps/sig-apps/0032-portable-service-definitions.md

```diff
@@ -0,0 +1,143 @@
+approvers:
+  - "@mattfarina"
+  - "@prydonius"
+  - "@kow3ns"
+editor: TBD
+creation-date: 2018-11-13
+last-updated: 2018-11-13
+status: provisional
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# Portable Service Definitions
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a KEP and for highlighting any additional information provided beyond the standard KEP template.
+[Tools for generating][] a table of contents from markdown are available.
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories [optional]](#user-stories-optional)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+## Summary
+
+The goal of this feature is to enable an application to be deployed into multiple environments while relying on external services that are not part of the application. The service instances may or may not be managed by Kubernetes. For example, take a WordPress application that relies on MySQL. If the application is running in GKE or AKS it may leverage a Google or Azure MySQL as a Service. If it is running in Minikube or in a bare metal cluster it may use MySQL managed by the Kubernetes cluster. In all of these cases, a resource is declared asking for the service (e.g., MySQL) and the credentials to use the service are returned in a common manner. This enables portability from one environment to another.
+
+## Motivation
+
+Workload portability is commonly cited as a goal for those deploying workloads. The Kubernetes API can provide a common API for deploying workloads across varying environments enabling some level of portability. For example, a deployment can be run in a cluster running on GKE, AKS, EKS, or clusters running elsewhere.
+
+But, many applications rely on software as a service (SaaS). The reason for this is to push the operational details on to someone else who specializes in that particular service so the application developers and operators can focus on their application and business logic.
+
+Currently, service portability as declared through Kubernetes with a common API is not possible. This feature set would provide a means to solve that.
+
+### Goals
+
+* Provide a common way to request common services (e.g., MySQL)
+* Provide a common means to obtain credentials to use the service
+* Provide a common method to detect which services are available
+* Provide a system that can be implemented for the major public clouds, on-premise clusters, and local cluster (e.g., Docker for Mac)
```

Comment:
What's the scope of that "system"? Isn't the implementation vendor-specific aka "operators"?

Reply:
I will try to come up with a better word here. A system is not a solution nor is it a product. There word is too generic and we can read different things from it. Thanks for catching that.

Reply:
Could this be a spec, or does this go further than that? Framework?